File: solutions/java/src/main/java/com/epi/CanStringBePalindromeSorting.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Arrays;

public class CanStringBePalindromeSorting {
  // @include
  public static boolean canFormPalindrome(String s) {
    char[] a = s.toCharArray();
    Arrays.sort(a);
    int oddCount = 0;
    int numCurrChar = 1;

    for (int i = 1; i < a.length && oddCount <= 1; ++i) {
      if (a[i] != a[i - 1]) {
        if ((numCurrChar % 2) != 0) {
          ++oddCount;
        }
        numCurrChar = 1;
      } else {
        ++numCurrChar;
      }
    }
    if ((numCurrChar & 1) != 0) {
      ++oddCount;
    }

    // A string can be permuted as a palindrome if the number of odd time
    // chars <= 1.
    return oddCount <= 1;
  }
  // @exclude
}

File: solutions/python/can_string_be_palindrome_sorting.py
import itertools


# @include
def can_string_be_a_palindrome(s):
    s = ''.join(sorted(s))
    odd_count, num_curr_char = 0, 1

    for i in range(1, len(s)):
        if s[i] != s[i - 1]:
            if num_curr_char & 1:
                odd_count += 1
            num_curr_char = 1
            if odd_count > 1:
                break
        else:
            num_curr_char += 1
    if num_curr_char & 1:
        odd_count += 1

    # A string can be permuted as a palindrome if the number of odd time chars
    # <= 1.
    return odd_count <= 1
# @exclude


def can_string_be_a_palindrome_pythonic(s):
    return sum(len(list(group)) & 1 for _,
               group in itertools.groupby(sorted(s))) <= 1
